core/jobs/src/main/java/com/cognifide/aet/job/common/comparators/source/SourceComparator.java

```diff
@@ -16,164 +16,109 @@
-
-        dataSource = dataFilterJob.modifyData(dataSource);
-
-        LOGGER.info("Successfully ended data modifications using  {}. Company: {} Project: {}",
-            dataFilterJob.getInfo(), properties.getCompany(), properties.getProject());
-
-        patternSource = dataFilterJob.modifyPattern(patternSource);
-
-        LOGGER.info("Successfully ended pattern modifications using {}. Company: {} Project: {}",
-            dataFilterJob.getInfo(), properties.getCompany(), properties.getProject());
-      }
-
-      if (StringUtils.isNotBlank(patternSource)) {
-        boolean compareTrimmedLines = shouldCompareTrimmedLines(sourceCompareType);
-        final List<ResultDelta> deltas = diffParser
-            .generateDiffs(patternSource, dataSource, compareTrimmedLines);
-        if (deltas.isEmpty()) {
-          result = new ComparatorStepResult(null, ComparatorStepResult.Status.PASSED, false);
-        } else {
-          result = new ComparatorStepResult(artifactsDAO.saveArtifactInJsonFormat(properties,
-              Collections.singletonMap("differences", deltas)),
-              ComparatorStepResult.Status.FAILED, true);
-          result.addData("formattedPattern", artifactsDAO.saveArtifact(properties, patternSource));
-          result.addData("formattedSource", artifactsDAO.saveArtifact(properties, dataSource));
-          result.addData("sourceCompareType", sourceCompareType.name());
-        }
-      } else {
-        result = new ComparatorStepResult(null, ComparatorStepResult.Status.PASSED);
-      }
-    } catch (Exception e) {
-      throw new ProcessingException(e.getMessage(), e);
+      return compareSources();
+    } catch (Exception ex) {
+      throw new ProcessingException(ex.getMessage(), ex);
     }
-
-    return result;
   }
 
+  @SuppressWarnings("unchecked")
+  private ComparatorStepResult compareSources() throws IOException, ProcessingException {
+    sources.generate(sourceCompareType);
+    return getResultOfCompare();
+  }
 
-  private boolean shouldCompareTrimmedLines(SourceCompareType sourceCompareType) {
-    return SourceCompareType.ALLFORMATTED.equals(sourceCompareType)
-        || SourceCompareType.MARKUP.equals(sourceCompareType);
+  private ComparatorStepResult getResultOfCompare() {
+    List<ResultDelta> deltas = calculateDeltas();
+    ComparatorStepResult.Status status = calculateStatus(deltas);
+    ComparatorStepResult result = createNewStepResult(deltas, status);
+    addFormattedSources(result);
+    return result;
   }
 
-  @Override
-  public void setParameters(final Map<String, String> params) throws ParametersException {
-    if (params.containsKey(SOURCE_COMPARE_TYPE)) {
-      this.sourceCompareType = SourceCompareType
-          .valueOf(params.get(SOURCE_COMPARE_TYPE).toUpperCase());
-    }
+  private List<ResultDelta> calculateDeltas() {
+    return Optional.ofNullable(sources.getPatternSource())
+        .filter(StringUtils::isNotBlank)
```

Comment:
This filter - does it mean, that if pattern is blank, then we always get empty list from this method? That is not good...

What if we accepted blank page as a pattern? We would like to get information when the page changed to not blank. Could we have some unit test for pattern that is blank?

Reply:
Original code:

```
if (StringUtils.isNotBlank(patternSource)) {
        boolean compareTrimmedLines = shouldCompareTrimmedLines(sourceCompareType);
        deltas = diffParser
            .generateDiffs(patternSource, dataSource, compareTrimmedLines);
        if (deltas.stream().anyMatch(d -> d.getType().equals(TYPE.CHANGE))) {
          status = Status.FAILED;
        } else {
          status = Status.PASSED;
        }
} else {
    status = Status.PASSED;
}
```
I will change it to `notNull` and add some test with blank pattern